### ISSUE TITLE
Layout of graphs by component and merge the results

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1831]
**What is the feature or improvement you would like to see?**

I would like to see better support for layout out each connected component of graphs separately, and then having various methods to merge the results back together.

Mathematica has this feature, and it is very useful. To see how it works, go to the [GraphLayout](https://reference.wolfram.com/language/ref/GraphLayout.html) doc page -> Scope -> Packing Layouts, and see the examples for each layout. See also [PackingMethod](https://reference.wolfram.com/language/ref/PackingMethod.html). I think there's no reason not to borrow from a design that works well.

Possible technical implementation:

 - Have a driver function that does the component decomposition, separate layouts, and merging, all in one go. How do we easily support all existing layouts, as well as multiple packing layouts, in a single function?
 - Have a number of merging methods.
 - Have at least one method that requests "no decomposition".
 - Have control over how components are sorted before merging them together (for example, larger graphs on top, small on bottom).

**Use cases for the feature**

 - Some layout methods that igraph have simply do not work well for disconnected graphs. Those that work do not always produce nice results.
 - Flexibility in merging components is various useful ways.

**References**

 - Mathematica's "Tech Note" in its documentation: https://reference.wolfram.com/language/tutorial/GraphDrawing.html, and references within:
 - Freivalds, K., U. Dogrusoz, and P. Kikusts, "Disconnected Graph Layout and the Polyomino Packing Approach." Lecture Notes in Computer Science: Revised Papers from the 9th International Symposium on Graph Drawing 2265 (2001): 378–391.
 - Word-cloud type packing implemented by @szhorvat: https://mathematica.stackexchange.com/a/2529/12

TODO: Add more.

**Comment by szhorvat** [comment]
@ntamas I think it would be nice to think a little bit about the design of this feature, just in case we discover that it would be useful [to break the API](https://github.com/igraph/igraph/wiki/Planned-API-changes-for-0.10) in some way (e.g. make the interface of layout functions more uniform, so that they can easily be used from the same driver). This is why I posted this wishlist item now.

**Comment by ntamas** [comment]
I agree with the direction of this proposal in general; having a default strategy for packing component layouts next to each other would allow the implementation of each layout to focus on the complicated part instead, i.e. the layout of a single connected component.

However, I'm afraid that we can't really create an easy-to-use generic driver function due to the limitations of C; the driver function should be able to call any _abstract_ layout algorithm without knowing what parameters the layout algorithm takes, so the best I could come up with is something like this:

First, we need a data structure to store layout algorithm parameters. For sake of simplicity, I now assume that layout algorithms take only doubles as parameters:

```
typedef struct {
    const char* name;
    igraph_real_t value;
} igraph_layout_parameter_t;
```

Then, we need a special sentinel value to mark the end of a layout parameter array:

```
#define IGRAPH_NO_MORE_LAYOUT_PARAMETERS { 0, 0 }
```

And then we can create a typedef for a generic layout algorithm:

```
typedef igraph_error_t (*igraph_layout_func_t)(const igraph_t*, igraph_matrix_t*, const igraph_vector_int_t*, const igraph_layout_parameter_t*)
```

where the first argument is the graph to lay out, the second argument is an output argument for the layout, the third argument is the list of vertex IDs to consider in the layout (to allow us to lay out subcomponents without actually constructing the subgraphs), and the fourth argument is the list of parameters to pass to the actual layout algorithm.

And then we need to refactor all our layout algorithms to at least have an _internal_ implementation that conforms to the `igraph_layout_func_t` typedef. And then we also need helper functions for the layout algorithms to find an arbitrary parameter with a given name in the array of layout parameters, returning a default value if the layout parameter does not exist.

It seems like this would make the layout API less friendly in the C layer, but maybe no one expects the C API to be user-friendly anyway :)

**Comment by vtraag** [comment]
Instead of creating a separate API for ensuring that a layout function can be uniformly called for each separate component, would it not make more sense to implement simply a packing layout that takes a layout for each component and then tries to pack this tightly? People would hence be required to call a layout function themselves for each component, but can then easily pack all the layouts using such a layout packing function.
